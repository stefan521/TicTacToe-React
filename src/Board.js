import React from 'react';
import Square from './Square.js';

class Board extends React.Component {
  renderSquare(squareNumber) {
    return (
       <Square
          value={this.props.squares[squareNumber]}
          onClick={() => this.props.onClick(squareNumber)}
          key={squareNumber}
       />
    );
  }

  render() {
    let rows = [];
    let boardSize = 3;

    for(let i = 0; i < boardSize; i++) {
      let row = [];
      for(let j = 0; j < boardSize; j++) {
        let squareNumber = i * boardSize + j;
        row.push(this.renderSquare(squareNumber));
      }
      rows.push(<div className="board-row" key={i}>{row}</div>);
    }

    return (<div>{rows}</div>);
  }
}

export default Board;
