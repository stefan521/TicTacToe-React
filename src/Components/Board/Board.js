import React from 'react';
import Square from '../Square/Square.js';

class Board extends React.Component {
  pickSquareColor(squareNumber) {
    let winLine = this.props.winLine;
    if(winLine && winLine.includes(squareNumber)){
      return "lightgreen";
    } else {
      return "white";
    }
  }

  renderSquare(squareNumber) {
    return (
       <Square
          value={this.props.squares[squareNumber]}
          onClick={() => this.props.onClick(squareNumber)}
          color={this.pickSquareColor(squareNumber)}
          key={squareNumber}
       />
    );
  }

  render() {
    let rows = [];
    let boardSize = 3;

    for(let i = 0; i < boardSize; i++) {
      let row = [];
      for(let j = 0; j < boardSize; j++) {
        let squareNumber = i * boardSize + j;
        row.push(this.renderSquare(squareNumber));
      }
      rows.push(<div className="board-row" key={i}>{row}</div>);
    }

    return (<div>{rows}</div>);
  }
}

export default Board;
