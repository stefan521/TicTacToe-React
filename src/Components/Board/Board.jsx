// @flow
import React from 'react';
import Square from '../Square/Square.jsx';

type Props = {
  winLine: Array<number>,
  squares: Array<string>,
  onClick: Function
};

type State = {
  boardSize: number;
};

class Board extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      boardSize: 9,
    };
  }

  isWinSquare(squareNumber: number) {
    const { winLine } = this.props;
    if (winLine && winLine.includes(squareNumber)) {
      return true;
    }

    return false;
  }

  renderSquare(squareNumber: number) {
    const { squares } = this.props;
    const { onClick } = this.props;

    return (
      <Square
        value={squares[squareNumber]}
        onClick={() => onClick(squareNumber)}
        winSquare={this.isWinSquare(squareNumber)}
        key={squareNumber}
      />
    );
  }

  render() {
    // eslint-disable-next-line prefer-const
    let squares = Array(9);
    const { boardSize } = this.state;

    for (let i = 0; i < boardSize; i += 1) {
      squares[i] = this.renderSquare(i);
    }

    return (
      <div className="board-wrapper">
        <div className="game-board">{squares}</div>
      </div>
    );
  }
}

export default Board;
