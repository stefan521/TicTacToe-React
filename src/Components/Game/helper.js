function checkWinner(lines, squares) {
  for (let i = 0; i < lines.length; i += 1) {
    const [a, b, c] = lines[i];

    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
      return {
        gameEnded: true,
        winner: squares[a],
        line: lines[i],
      };
    }
  }

  return null;
}

function isBoardFull(lines, squares) {
  if (!squares.includes(null)) {
    return {
      gameEnded: true,
      winner: null,
      line: null,
    };
  }

  return null;
}

function getGameStatus(squares) {
  const lines = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6],
  ];

  const winner = checkWinner(lines, squares);
  if (winner) {
    return winner;
  }

  const draw = isBoardFull(lines, squares);
  if (draw) {
    return draw;
  }

  return {
    gameEnded: false,
    winner: null,
    line: null,
  };
}

// eslint-disable-next-line import/prefer-default-export
export { getGameStatus };
