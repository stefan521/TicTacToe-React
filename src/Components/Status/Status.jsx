// @flow
import React from 'react';

type Props = {
  status: {
    gameEnded: boolean,
    winner: string
  },
  xIsNext : boolean
};

function getStatus(status, xIsNext) {
  let statusLine;

  if (status.gameEnded && status.winner) {
    statusLine = `Winner: ${status.winner}`;
  } else if (status.gameEnded) {
    statusLine = 'Draw';
  } else {
    statusLine = `Next player: ${xIsNext === true ? 'X' : 'O'}`;
  }

  return statusLine;
}

function Status(props: Props) {
  const { status } = props;
  const { xIsNext } = props;

  return (
    <div className="game-status">
      {getStatus(status, xIsNext)}
    </div>
  );
}

export default Status;
