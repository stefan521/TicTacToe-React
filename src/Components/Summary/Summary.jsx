/* eslint-disable class-methods-use-this */
import React from 'react';

class Summary extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      descending: true,
    };
  }

  getPlaySummary(move) {
    if (move.player === null) {
      return 'go to game start';
    }

    const summary = `go to move ${move.player} at `;
    const x = move.square % 3;
    let y = Math.floor(move.square / 3);
    // the origin of the xy axis is top left, not bottom left
    y = 2 - y;
    const coordinates = `(${x}, ${y})`;
    return (summary + coordinates);
  }

  getMoveAtIndex(index) {
    const { descending } = this.state;
    const { moves } = this.props;

    if (descending) {
      return index;
    }
    const reversedIndex = moves - index - 1;
    return reversedIndex;
  }

  getMoves() {
    const { descending } = this.state;
    const { moves } = this.props;
    const { onClick } = this.props;
    const { stepNumber } = this.props;

    const orderedMoves = descending ? moves : moves.slice().reverse();
    // make a li for each move
    return orderedMoves.map((move, index) => {
      const summary = this.getPlaySummary(move);
      return (
        // eslint-disable-next-line react/no-array-index-key
        <li key={index}>
          <button
            type="button"
            onClick={() => onClick(this.getMoveAtIndex(index))}
            style={(this.getMoveAtIndex(index) === stepNumber)
              ? { fontWeight: 'bold' } : { fontWeight: 'normal' }}
          >
            {summary}
          </button>
        </li>
      );
    });
  }

  reverseMoves() {
    const { descending } = this.state;

    this.setState({
      descending: !descending,
    });
  }

  render() {
    const { descending } = this.state;

    return (
      <div>
        <button
          type="button"
          onClick={() => this.reverseMoves()}
        >
          { `view moves ${descending ? 'descending' : 'ascending'}` }
        </button>
        <ol reversed={!descending}>
          {this.getMoves()}
        </ol>
      </div>
    );
  }
}

export default Summary;
