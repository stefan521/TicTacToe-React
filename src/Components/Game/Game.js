import React from 'react';
import Board from '../Board/Board.js';
import Summary from '../Summary/Summary.js';
import Status from '../Status/Status.js';
import { getGameStatus } from './helper.js';

class Game extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      history: [{
        squares: Array(9).fill(null),
      }],
      moves: [{
        player: null,
        square: null,
      }],
      stepNumber: 0,
      xIsNext: true,
    }
  }

  handleClick(squareNumber) {
    // first object in the history is the initial empty board as an array
    const history = this.state.history.slice(0, this.state.stepNumber + 1);
    // the first move when the board is empty is null, null too
    const moves = this.state.moves.slice(0, this.state.stepNumber + 1);
    const current = history[history.length - 1];
    const squares = current.squares.slice();
    const player = this.state.xIsNext ? 'X' : 'O';

    if (getGameStatus(squares).gameEnded || squares[squareNumber]) {
      return;
    }

    squares[squareNumber] = player;

    this.setState({
      history: history.concat([{
        squares: squares,
      }]),
      moves: moves.concat({player: player, square: squareNumber}),
      stepNumber: history.length,
      xIsNext: !this.state.xIsNext,
    });
  }

  jumpTo(stepNumber) {
    this.setState({
      stepNumber: stepNumber,
      xIsNext: (stepNumber % 2) === 0,
    });
  }

  render() {
    const history = this.state.history;
    const current = history[this.state.stepNumber];
    const status = getGameStatus(current.squares);

    return (
      <div className="game">
        <div className="game-board">
          <Board
            squares={current.squares}
            onClick={(square) => this.handleClick(square)}
            winLine ={status.line}
          />
        </div>
        <div className="game-info">
          <Status
            xIsNext={this.state.xIsNext}
            status={status}
          />
          <Summary
            moves={this.state.moves}
            stepNumber={this.state.stepNumber}
            onClick={(step) => this.jumpTo(step)}
          />
        </div>
      </div>
    );
  }
}

export default Game;
