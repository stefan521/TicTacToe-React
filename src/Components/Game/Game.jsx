// @flow
import React from 'react';
import Board from '../Board/Board.jsx';
import Summary from '../Summary/Summary.jsx';
import { getGameStatus } from './helper';

type Props = any;

type State = {
  history: Array<any>,
  moves: Array<any>,
  stepNumber: number,
  xIsNext: boolean
};

class Game extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      history: [{
        squares: Array(9).fill(null),
      }],
      moves: [{
        player: null,
        square: null,
      }],
      stepNumber: 0,
      xIsNext: true,
    };
  }

  handleClick(squareNumber: number) {
    const { history } = this.state;
    const { stepNumber } = this.state;
    const { xIsNext } = this.state;
    const { moves } = this.state;

    // first object in the history is the initial empty board as an array
    const historyUntilStep = history.slice(0, stepNumber + 1);
    // the first move when the board is empty is null, null too
    const movesUntilStep = moves.slice(0, stepNumber + 1);
    const current = historyUntilStep[historyUntilStep.length - 1];
    const squares = current.squares.slice();
    const player = xIsNext ? 'X' : 'O';

    if (getGameStatus(squares).gameEnded || squares[squareNumber]) {
      return;
    }

    squares[squareNumber] = player;

    this.setState({
      history: historyUntilStep.concat([{
        squares,
      }]),
      moves: movesUntilStep.concat({ player, square: squareNumber }),
      stepNumber: historyUntilStep.length,
      xIsNext: !xIsNext,
    });
  }

  jumpTo(stepNumber: number) {
    this.setState({
      stepNumber,
      xIsNext: (stepNumber % 2) === 0,
    });
  }

  render() {
    const { history } = this.state;
    const { stepNumber } = this.state;
    const { xIsNext } = this.state;
    const { moves } = this.state;

    const current = history[stepNumber];
    const status = getGameStatus(current.squares);

    return (
      <div className="game">
        <Board
          squares={current.squares}
          onClick={(square) => this.handleClick(square)}
          winLine={status.line}
        />
        <Summary
          moves={moves}
          stepNumber={stepNumber}
          onClick={(step) => this.jumpTo(step)}
          status={status}
          xIsNext={xIsNext}
        />
      </div>
    );
  }
}

export default Game;
