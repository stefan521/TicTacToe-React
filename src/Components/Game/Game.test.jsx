import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import Game from './Game.jsx';
import Board from '../Board/Board.jsx';
import Status from '../Status/Status.jsx';
import Summary from '../Summary/Summary.jsx';

describe('Game', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Game />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('hasn\'t changed structure', () => {
    const game = <Game />;

    expect(game).toMatchSnapshot();
  });

  it('has correct components inside', () => {
    const wrapper = shallow(<Game />);

    expect(wrapper.find(Board)).toHaveLength(1);
    expect(wrapper.find(Summary)).toHaveLength(1);
    expect(wrapper.find(Status)).toHaveLength(1);
  });
});
