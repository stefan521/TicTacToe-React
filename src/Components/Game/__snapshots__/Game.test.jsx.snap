// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game hasn't changed structure 1`] = `<Game />`;
